cmd/migrate_from_pinecone.go

```diff
@@ -0,0 +1,326 @@
+	if err != nil {
+		return fmt.Errorf("failed to count points in source: %w", err)
+	}
+
+	err = r.prepareTargetCollection(ctx, sourceClient, targetClient)
+	if err != nil {
+		return fmt.Errorf("error preparing target collection: %w", err)
+	}
+
+	displayMigrationStart("pinecone", r.Pinecone.IndexName, r.Qdrant.Collection)
+
+	err = r.migrateData(ctx, sourceIndexConn, targetClient, sourcePointCount)
+	if err != nil {
+		return fmt.Errorf("failed to migrate data: %w", err)
+	}
+
+	targetPointCount, err := targetClient.Count(ctx, &qdrant.CountPoints{
+		CollectionName: r.Qdrant.Collection,
+		Exact:          qdrant.PtrOf(true),
+	})
+	if err != nil {
+		return fmt.Errorf("failed to count points in target: %w", err)
+	}
+
+	pterm.Info.Printfln("Target collection has %d points\n", targetPointCount)
+
+	return nil
+}
+
+func (r *MigrateFromPineconeCmd) connectToPinecone() (*pinecone.Client, *pinecone.IndexConnection, error) {
+	client, err := pinecone.NewClient(pinecone.NewClientParams{
+		ApiKey: r.Pinecone.APIKey,
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create Pinecone client: %w", err)
+	}
+
+	indexConn, err := client.Index(pinecone.NewIndexConnParams{
+		Host:      r.Pinecone.Host,
+		Namespace: r.Pinecone.Namespace,
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to connect to Pinecone index: %w", err)
+	}
+
+	return client, indexConn, nil
+}
+
+func (r *MigrateFromPineconeCmd) countPineconeVectors(ctx context.Context, indexConn *pinecone.IndexConnection) (uint64, error) {
+	stats, err := indexConn.DescribeIndexStats(ctx)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get index statistics: %w", err)
+	}
+
+	return uint64(stats.TotalVectorCount), nil
+}
+
+func (r *MigrateFromPineconeCmd) prepareTargetCollection(ctx context.Context, sourceClient *pinecone.Client, targetClient *qdrant.Client) error {
+	if !r.Migration.CreateCollection {
+		return nil
+	}
+
+	targetCollectionExists, err := targetClient.CollectionExists(ctx, r.Qdrant.Collection)
+	if err != nil {
+		return fmt.Errorf("failed to check if collection exists: %w", err)
+	}
+
+	if targetCollectionExists {
+		pterm.Info.Printfln("Target collection '%s' already exists. Skipping creation.", r.Qdrant.Collection)
+		return nil
+	}
+
+	indexes, err := sourceClient.ListIndexes(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to list Pinecone indexes: %w", err)
+	}
+
+	var foundIndex *pinecone.Index
+	for i := range indexes {
+		if indexes[i].Name == r.Pinecone.IndexName {
+			foundIndex = indexes[i]
+			break
+		}
+	}
+
+	if foundIndex == nil {
+		return fmt.Errorf("index %q not found in Pinecone", r.Pinecone.IndexName)
+	}
+
+	distanceMapping := map[pinecone.IndexMetric]qdrant.Distance{
+		pinecone.Cosine:     qdrant.Distance_Cosine,
+		pinecone.Euclidean:  qdrant.Distance_Euclid,
+		pinecone.Dotproduct: qdrant.Distance_Dot,
+	}
+
+	var createReq *qdrant.CreateCollection
+
+	switch foundIndex.VectorType {
```

Comment:
In a Pinecone index, you can either store dense vectors or sparse vectors. Not both.